src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -2712,7 +2726,17 @@ public void visitForLoop(JCForLoop tree) {
                     uninitsSkip.assign(uninits);
                     uninitsSkip.inclRange(firstadr, nextadr);
                 }
-                scan(tree.body);
+                final Bits reassignedInForLoopBodyPrev = reassignedInForLoopBody;
+                reassignedInForLoopBody = new Bits();
+                try {
+                    scan(tree.body);
+                    for (int adr = forLoopVarMin; adr < forLoopVarMax; adr++) {
```

Comment:
Question: since we already have a dedicated flag for loop variables - could we set the flag on _all_ loop variables (before the body), and then remove it in `letInit` (same as we do for effectively final flag) ? This way, at the end of the loop, the variables that still have the flag set will be the ones that have not been reassigned, correct?

Reply:
> * We would need a new flag that says "we are currently in the for() loop body" and not, for example, in the step, where we should leave the flag alone.

Well, my idea for doing that was to just set the flag on the loop variables before scanning the loop body (as you seem to do in the patch you show)

> 
> * We would also need to ensure we don't get confused with nested `for()` loops. So the aforementioned flag can't just be a boolean (because otherwise which `for()` loop are we talking about?). For example, a symbol address range instead.

This I don't get. If you have a loop nested inside another loop, you want assignments in the inner loop to affect whether the outer loop variable is considered a "loop capturable variable" or not. So I'm not sure - it seems to me `letInit` should happily remove flags for any variable it sees marked with `FOR_LOOP_BODY_MAY_CAPTURE` (as that is surely a variable of a for loop and we must be in that loop's body for the variable to have that flag set?)

Reply:
> So I'm not sure - it seems to me `letInit` should happily remove flags for any variable it sees marked with `FOR_LOOP_BODY_MAY_CAPTURE` (as that is surely a variable of a for loop and we must be in that loop's body for the variable to have that flag set?)

Concretely, why not, in `letInit`:

```
// In basic for() loop bodies, variables that are reassigned can no longer be captured
if (sym.flags_field & FOR_LOOP_BODY_MAY_CAPUTURE && !uninits.isMember(sym.adr)) {
    sym.flags_field &= ~FOR_LOOP_BODY_MAY_CAPTURE;
}
```

Reply:
> This I don't get. If you have a loop nested inside another loop, you want assignments in the inner loop to affect whether the outer loop variable is considered a "loop capturable variable" or not. So I'm not sure - it seems to me `letInit` should happily remove flags for any variable it sees marked with `FOR_LOOP_BODY_MAY_CAPTURE` (as that is surely a variable of a for loop and we must be in that loop's body for the variable to have that flag set?)

I think you are correct, but I think we're referring to different flags. I was referring to this flag:

> We would need a new flag that says "we are currently in the for() loop body"

That's would be a flag in `AssignAnalyzer`, not `Symbol`. Sorry for the mixing of terminology.

To be clear, here's the (first) problem I was worried about:
```java
for (int i = 1; i <= 3; i++) {
    for (int j = 1; j <= 3; j++) {
        Runnable r = () -> System.out.println(j);   // false error here
    }
}
```
Explanation: We set the boolean flag `WeAreCurrentlyInTheForLoopBody` just before entering the outer loop body. Then when entering the inner loop step (which occurs prior entering to the inner loop body), the flag is still set, and so then we would incorrectly mark `j` has being reassigned in the body - but it's the wrong body that the flag is referring to. We're in the outer loop body but not the inner loop body.

This could be addressed by a stash/pop bracketing of the `WeAreCurrentlyInTheForLoopBody` flag around the _entire_ `visitForLoop()` method instead of only around the recursion into the body.

But then you still have a second problem, as shown in this example:

```java
for (int i = 1; i <= 3; i++) {
    for (int j = 1; j <= 3; i++, j++) {
        Runnable r = () -> System.out.println(i);
    }
}
```
Now the `i++` that is the inner loop step does not invalidate the capture of `i` as it should, because `WeAreCurrentlyInTheForLoopBody` will be false.  So in summary `WeAreCurrentlyInTheForLoopBody` can't just be a boolean flag because: Which loop are we talking about?

Reply:
> Concretely, why not, in `letInit`:

Because `visitForLoop()` itself has a loop and recurses into the parts of the `for()` statement multiple times. So when you loop back to visit `step` after visiting `body` for the first time, the flag would get reset.

So then you say, "Let's set a flag that let's this code know that we're in the _body_ of the loop" like this:
```java
// In basic for() loop bodies, variables that are reassigned can no longer be captured
if (inForLoopBody && (sym.flags_field & FOR_LOOP_BODY_MAY_CAPTURE) != 0 && !uninits.isMember(sym.adr)) {
    sym.flags_field &= ~FOR_LOOP_BODY_MAY_CAPTURE;
}
```
and then you realize a boolean flag is not sufficient, and then ... well, see previous discussion :)

Another probem is that after the first loop, the variable is no longer DU...

Reply:
> Because `visitForLoop()` itself has a loop and recurses into the parts of the `for()` statement multiple times.

I believe this was the source of the misunderstanding. The `scan(tree.body)` in `AssignAnalyzer` is indeed buried in a loop as we need a couple of passes to make sure to detect bad cases of multiple assignments e.g. to the same blank final variable.

Note that even in the current PR you `letInit` will be called twice, which in principle can be problematic in cases like:

```
for (int i ; ,..) {
    i = m();
    ...
}
```

If I understand correctly, the way the code in this PR works for the above is:
* in the first DA/DU pass of the loop body (`flowKind` == `NORMAL`), `i` is DU when it's assigned. So it is *not* added to the list of loop variable reassigned in body
* at the end of the pass, we see that `i` is not reassigned, so we augment it with the `FOR_LOOP_BODY_MAY_CAPTURE` flag
* then we do a second DA/DU pass of the loop body (`flowKind` == `SPECULATIVE`), which calls `letInit` again. This time `i` is not DU when assigned, so `i` ends up in the list of reassigned loop variables.
* at the end of the pass we do nothing (since we see `i` as reassigned). Luckily, the flag set in the first DA/DU pass is still there, and will be used by `CaptuerAnalyzer`.

This begs the question: should `letInit` even bother to deal with reassigned loop variables in the speculative DA/DU pass? Seems to me that pass will generally cause spurious results because of the changed DA/DU status of the variables involved.

Regardless, I think I found an issue with the current implementation when there are nested loops:

```
    void test() {
        for (int i = 0; ; ) {
            for (; ;) {
                i = 42;
                Runnable r = () -> System.out.println(i);
            }
        }
    }
```

This compiles fine, even though `i` is reassigned when DA. The problem seems to be that we lose track of the reassigned bit set (e.g. the inner loop will overwrite that), and since the variable is reassigned from the inner loop, that is problematic. Removing the inner loop results in a compile-time error, as expected.

We should either track "active" loop variables in a list when we append/remove from. Or, we should use the various Bits::incl/excl methods. E.g. flow should not create a new bit set on each loop, but augment the existing bit set (the initial could be the "all zero" set) with the bits corresponding to the variables of the loop under consideration. Then, once we have finished inspecting a loop, we "exclude" its variable bits from the sets before returning from the visitor. This way the "outer" loop will be able to "see" reassignments happening in inner loops.

Reply:
> Regardless, I think I found an issue with the current implementation when there are nested loops:
> 
> ```
>     void test() {
>         for (int i = 0; ; ) {
>             for (; ;) {
>                 i = 42;
>                 Runnable r = () -> System.out.println(i);
>             }
>         }
>     }
> ```

Issues aside, I think a related example:

```java
void test() {
        for (int i ; ; ) {
            for (; ;) {
                i = 42;
                Runnable r = () -> System.out.println(i);
            }
        }
    }
```
Poses an interesting question: should `i` be capturable from the PoV of the inner loop? After all, in the inner loop we know there one value for `i` on each iteration.

Relatedly, what about this?

```
void test(int x) {
        int i = 42;
        if (x > 42) {
            i = 43;
        } else {
            Runnable r = () -> System.out.println(i);
        }
    }
```

We know that `x` is not reassigned in the `else` branch. But the analysis we have now is not sophisticated enough to detect that.

I think there seem to be some more general principle here that is struggling to get out: sometimes there might be region of code where we know a variable can hold only one value (even if in other regions that variable might be reassigned multiple times). In such case you can imagine to take the content of the variable and assign it to a _new_ variable, local to that "well-behaved" region. That new variable could, of course, be captured. The loop case (handled by this patch) seems to be a special case of this situation.

Reply:
> Regardless, I think I found an issue with the current implementation when there are nested loops:

Yep - thanks, very interesting. To clarify the intent, a variable is capturable if it is not **reassigned in the loop**, which is defined as:

* Wherever the variable occurs as the left hand side in an assignment expression, it is definitely unassigned
* The variable never occurs as the operand of a prefix or postfix increment or decrement operator

In your example with `i = 42`, the variable `i` is not DU at that point (even though it takes two loops around the inner loop in `AssignAnalyzer` to figure that out). So according to the original intent, it should not be capturable.

I'll work on fixing this (and add a new test case).

> I think there seem to be some more general principle here that is struggling to get out...

Agreed... I had a similar vague thought while wondering "What about `while { }` and `do { }` loops?" In other words, is there some coherent way that we could we spread this new capturable goodness more widely?

One simple idea for generalizing the current approach: Within any scope _S_, any normal variable _v_ that is visible in _S_ is capturable if it is not reassigned in _S_. I don't know if the JLS actually defines "scope" but you know the idea: a region of code in which variables can be declared and be visible; so `{ ... }` curly braces obviously, but also the body of a `for()` statement.

An example that would then be allowed:
```java
int i, j = 0;
while (true) {
  j += i;
  {
    Runnable action = () -> foo(i + j);
    try {
        this.executor.submit(action).get();
    } catch (InterruptedException e) {
        break;
    }
  }
  i++;
}
```
Note the extra curly braces required to create the "well-behaved" region.

It's almost as if the "well-behaved" region "captures" the variables that it uses and makes them temporarily effectively final... which then begs the question, why not just cut out the middle man? That would then put us in a situation where any variable could be captured anywhere (which some have argued for), i.e., everything is effectively final at the point of capture.

So I'm not sure if a coherent middle ground exists or what it would be.

Reply:
> An example that would then be allowed:
> 
> ```java
> int i, j = 0;
> while (true) {
>   j += i;
>   {
>     Runnable action = () -> foo(i + j);
>     try {
>         this.executor.submit(action).get();
>     } catch (InterruptedException e) {
>         break;
>     }
>   }
>   i++;
> }
> ```

I tend to agree with this line of thinking. And this is IMHO, a sign that the design has not yet fully settled. The main thing to notice is that there's two stable design points:
1. only capture effectively final vars (where the language is today);
2. capture every variable - with the understanding that what's being captured is the _current_ value (and not a "mutable box" which keeps changing in real time - that would be awful).

Having special rules for loop variables is sending us down the slippery slope of finding the next stable point between (1) and (2) - assuming one exists. The risk here is that once we start relaxing things in some areas (e.g. loops) you can quickly run into issues where other deficiencies in capture become even more apparent.

Reply:
> Having special rules for loop variables is sending us down the slippery slope of finding the next stable point between (1) and (2) - assuming one exists. The risk here is that once we start relaxing things in some areas (e.g. loops) you can quickly run into issues where other deficiencies in capture become even more apparent.

I agree with all of that as well. There is one valid counter-argument though (reflected in the JEP draft) which is that we have already taken this current step with enhanced `for()` loops, so we're not really establishing any much of a new "frontier" with this feature by adding the same thing to basic `for()`. In other words, enhanced `for()` is an intermediate design point that's already been taken, for better or worse, so we might as well make the most of it. Or put more plaintively, "Please don't blame this feature for all the world's problems" :)

Regarding the bigger picture question, I'm skeptical that there exists any other design point between 1 and 2 that would be preferable to just jumping straight to 2, i.e., capture anything that's DA anywhere, but of course anything's possible.

Reply:
Thanks for the bug fix suggestion. Should be fixed in 00181f7f1b4.

Reply:
> It's almost as if the "well-behaved" region "captures" the variables that it uses and makes them temporarily effectively final... which then begs the question, why not just cut out the middle man? That would then put us in a situation where any variable could be captured anywhere (which some have argued for), i.e., everything is effectively final at the point of capture.

I am a long-time requester of this feature, in JDK-8300691.  I also considered whether other loops have a generalization, but gave it up after some thought.  I realized that the distinctive feature of old-for loops is their ability to bind variables that are not visible outside of the for loop, and also in the distinction between the region of the for loop header and for loop body.  Due to the traditional way these loops are used, it is very common (though not 100% universal) that such variables may be mutated in the header but are not mutated in the body.  The presence of two syntactic regions, one where variation is OK and one where variation is absent.  In the latter region, there is no benefit to telling the user "do not capture this because it might be different later on" (the rationale for non-capture of non-finals) because the only "later on" is another execution of the same loop body, in a different trip of the loop.

All this is to suggest why I think this request (and mine) should be _specifically and exactly only for old-for loops_, because they key off of specific and unique syntactic features (variable declaration + two syntax regions).

Therefore, I am pleading that we not confuse ourselves by attempting a generalization here.  Discarding the specifics of old-for loops in quest of some greater generalization is (IMO) a doomed effort, and even if it ultimately pans out, it should not stop us from dealing with the specifics of old-for loops.  If we are smart enough to come up with such a generalization, we will be smart enough to fit it into the current proposal, which is limited to the hardwired specifics of old-for loops.

I am the originator of the rule that forbids capture of non-finals (in Java 1.1).  I can tell you why it got there (if not how people think about it today).

The decision was made (after prototyping) _not to capture_ a full mutable state.  (…Such as a constant pointer to a one-element mutable array.)  There are numerous non-obvious reasons why that is a doomed decision. It's doomed even if good languages have done so — and as a Common Lisp user I saw the very sharp edges from the interaction between mutable variable capture and non-sequential executions the observe the mutable variables.  (Non-sequential includes saving a callback that reads the loop variable, and forking a thread in the loop.)

So, if we were not going to save access to the future states of the mutable variable, we had to take a snapshot.  That's nice, but then another problem appears:  How do you explain to the user that if he changes the value of the variable later on, a lambda made before won't see the change, while a lambda made after will see it?  Such a user will rightly think the language is keeping double copies of the variable, which a bad way to implement a variable.  The way out of the dilemma, in 1.1, was to insist that only final variables can be captured (and they must be definitely assigned with their one and only value).  This way you can have multiple copies under the hood, and there's no state to be shared, and the user's code will see a consistent binding for that variable.

It was not clear we could do better than that, but Java 8 made a step forward with the concept of "effectively final" variables, which are variables that "might as well" have final declaration; such variables can be captured by lambdas as well.

What we don't cover, now, are codes like either of these two examples:

```
// Ex 1: multiple assignments before the variable stabilizes
int K = 1;
if (p())  K = 2;
//FIX:  effectively-final int K = p() ? 2 : 1;
return () -> K;  //BAD cannot be effectively final
```

Here, you might be able fix the problem by refactoring the final so that it is defined by one expression, or defined by disjoint (nonoverlapping) assignments from a blank state.  (Blank finals were introduced by us in 1.1 as well, to open up more natural usages of final locals and final instance variables.)

```
final int K;  // blank
if (p())  K = 2;
else  K = 1;
return () -> K;  //OK
```

Second example:

```
// Ex 2: re-assignment where the user expects two values from one var
int V = 1;
Globals.T = new Thread(() -> V);  //will not end well

//maybe: if (p()) V = 2;
V = 2;  // update V for next lambda
return () -> V;  //BAD cannot be effectively final

//FIX:  split V into V1 and V2
```

It is clear that the V has two disjoint episodes, which could be renamed V1 and V2.  Could there be a predictable, understandable rule which would perform this split automatically (when a lambda needs it)?  Maybe, but it probably becomes difficult to do this automatically in all cases; some code will allow V to split cleanly, and other code won't.  It's not clear there is a viable user model here, that makes the splitting predictable and intuitive.

In the end, I doubt that the above examples can be made fit for lambdas without manual intervention (magic "implicitly split" or "implicitly disjointly initialized" features look unlikely to me).  But it might be possible to give users an explicit way to "lock down" a variable which is not yet final, so that (hereafter) it becomes subsequently final.  If the "lock-down" is limited to an enclosing block you might be able to salvage both of the examples above.

The feature I'm talking about (and maybe some of you may remember me talking about this before) is a re-declaration of a previously declared variable as (now) being final.  Let's call it "subsequently-final":


```
// Ex 1: multiple assignments before the variable stabilizes
int K = 1;
if (p())  K = 2;
subsequently-final K;
// sugar for final var K$ = K; 
// and with remaining K occurrences in current scope rewritten to K$
return () -> K;  //OK, rewritten to () -> K$

// Ex 2: re-assignment where the user expects two values from one var
int V = 1;

{  // the fix requires this bracket (so it cannot be fully automatic)
  subsequently-final V;
  // sugar for final var V$1 = V; 
  Globals.T = new Thread(() -> V);  // () -> V$1 instead
  // at end of scope, V$1 goes away, but V remains
}

//maybe: if (p()) V = 2;
V = 2;  // update V for next lambda
subsequently-final V;
// sugar for final var V$2 = V; 

return () -> V;  //OK, () -> V$2

//FIXED:  split V into V$1 and V$2
```

Such a subsequently-final feature provides an accounting (retroactive I trust) of how the current proposal works:

```
for (int N = 0, M = A.length; N < M; N++) {
   subsequently-final N;
   // capture N$ at will
}
```

The proposal here might be thought of "effectively-subsequently-final for variables declared by old-for loops".  In conclusion, it's a very specific patch to a commonly used syntax, which is reasonable by itself, and is unlikely to interfere with future changes to the language.

Reply:
> Thanks for the bug fix suggestion. Should be fixed in [00181f7](https://github.com/openjdk/jdk/commit/00181f7f1b458ab38efc8ffee933e97f5f07419a).

Looks good!

Reply:
> How do you explain to the user that if he changes the value of the variable later on, a lambda made before won't see the change, while a lambda made after will see it? Such a user will rightly think the language is keeping double copies of the variable, which a bad way to implement a variable. The way out of the dilemma, in 1.1, was to insist that only final variables can be captured (and they must be definitely assigned with their one and only value). This way you can have multiple copies under the hood, and there's no state to be shared, and the user's code will see a consistent binding for that variable.

I tend to agree with what you say here. To me though, I admit, the proposed new rules seem to push the language in the very direction that the careful design of the inner class mechanism wanted to avoid. I understand that you can think of an imperative, well-behaved loop induction variable behaving "as if" it was created fresh each time. And yet, according to the JLS, that variable is actually reassigned (case in point: you cannot make the lop variable `final`).

Overall, both the loop-only changes and your more general "subsequently-final" feature seem to defeat the very principles set out in the above: that allowing for multiple copies to be kept under the hood is confusing, and potentially surprising. For instance, the proposed changes allow this:

```
    record Box(int x) { }
    
    for (Box b = new Box(0) ; b.x < 10 ; b = new Box(b.x + 1)) {
            Runnable r = () -> System.out.println(b);
        }
    }
```
I think users will be surprised to realize that some "old" `Box` instances might be kept alive by `Runnable` (which might leak somewhere else).